### DISCUSSION TITLE
How to handle multiple pydantic  response in fast api

### DISCUSSION THREAD

**Discussion by gunjeetgulatigit** [fastapi/fastapi#7255] [Questions] [answered]
**Describe the bug**

When using Union on response_model, it doesn't return accurate result 
**To Reproduce**

###     Create three models

   ```
   from pydantic import BaseModel

      class parent(BaseModel):
          id: str

      class child1(parent):
         code: str
         user: Any

     class child2(parent):
       `error:` Any

class child3(parent):
    code: str
    user_data: Any

```

    Add it on response_model

```
     from typing import Union
    from fastapi import APIRouter

    router = APIRouter()

   @router.post('/verify', response_model=Union[child1, child3,child2])
   async def something(some: some):
      dict1 = {
          "id": 1,
          "code" : "213",
          "user":{'id':"123","pass":"pass"}
          }
      dict2 = {
          "id": 1,
          "error" : "213"
         }
       dict3 = {
          "id": 1,
          "code" : "213",
          "user_data":{'id':"123","pass":"pass"}
         }
       
     if cond1 :  
          return child1(**dict1).dict()
    elif cond2 :  
          return child2(**dict2).dict()
     else:
         return child3(**dict3).dict()
```

**output** 
**actual :**  
On running all the condition one by one output that i got is not correct 
when cond 1 gets true then 
```
      {
        "id": 1,
         "code" : "213",
         "user":{'id':"123","pass":"pass"}
        }
```
when cond 2 gets true then 
 ```
 {
        "id": 1,
         "error" : "213"
        }
```

when cond 3 gets true then
   ```
   {
        "id": 1,
         "code" : "213"
        }
```

Expected behavior

cond 1 and 2 output are correct 
**but cond3 is incorrect** 
i want cond3 output to have user_data field as below 
  
```
{
        "id": 1,
         "code" : "213"
       "user_data":{'id':"123","pass":"pass"}
        }

```
**how can i get the output to have all the fields of dict3**

i am new to fast api and can anyone help me with this issue, any resolution to this issue or any other way to get the
 **expected output with schema in swagger UI**

**Additional Info:** 
What i observed here is **Field type: Any** has issue , as when i specify the field to **str** or **dict** , it gives the correct output as expected

**Comment by ghost** [comment]
When you specify the response model in the decorator, eg: `response_model=Union[child1, child2, child3]` FastAPI Expects an Instance of that class, [See Here](https://fastapi.tiangolo.com/tutorial/response-model/) from the Documentation.

The line 
```py
return child3(**dict3).dict()
```

is what is causing the problem here. Do not convert it to a `dict`, FastAPI Will take care of that for you, 

What you need to be doing:

```py
return child3(**dict3)
```

**Comment by gunjeetgulatigit** [comment]
Hi @Rushyanth111 

`return child3(**dict3)`
Tried above statement to return the response but still getting the same output .

**Comment by ghost** [comment]
Ok, I tested this once again, and it seems to be true, the output @gunjeetgulatigit is getting is valid:

```json
{
  "id": "1",
  "code": "213",
  "user": null
}
``` 
When `return child3(**dict3)` is executed. 

I solved the problem by Introducing type-annotations into the pydantic Models.

So I converted your models a little:

```py
from typing import Dict

class parent(BaseModel):
    id: str


class child1(parent):
    code: str
    user: Dict[str, str] #Note here the Use of Dict[str, str]


class child2(parent):
    error: Any #Should be str, really, Using Any is not advisable. 


class child3(parent):
    code: str
    user_data: Dict[str, str] #Note here the Use of Dict[str, str]
```

Changing to the appropriate types ensured the output is what you expect. 

I'm a little confused as to how that happened, since the keys are obviously different, if anyone else could elucidate on that, I'd be happy.

**Comment by gunjeetgulatigit** [comment]
thanks @Rushyanth111 
Any idea to show default scheme in fast api doc and redoc for same part .
currently below response is displayed over fast-api doc
```
    Example Value
    Schema
  (no example available)
```

**Comment by ghost** [comment, marked as answer]
Currently, it is a known missing feature within the Swagger UI (See: #1083 ) and the generation of the example cannot be automatically done by Swagger UI, however it is possible to provide a manual example, which would mean that you have to make an example to be displayed.

[Link to the Relevant Issue Solution from #1083](https://github.com/tiangolo/fastapi/issues/1083#issuecomment-629736627)

[Link to Another Solution from #86](https://github.com/tiangolo/fastapi/issues/86#issuecomment-632412384)

The Example Shows you that you need to provide a manual example like so:

Create A Class that holds the Union of any of the three Classes:

```py

class parent(BaseModel):
    id: str


class child1(parent):
    code: str
    user: Dict[str, str]


class child2(parent):
    error: Any


class child3(parent):
    code: str
    user_data: Dict[str, str]


class UnionChild(BaseModel): # This will Hold the Union
    Child: Union[child1, child2, child3] #The Union Required

    class Config: #This is inside UnionChild, do not place it outside. 
        schema_extra = {
            "example": {                    #Mandatory field, this holds your example. Define Your Field from here. 
                "id": "1",
                "error": "Some Random String",
                "user": "OSAS UVUWE",
                "user_data": {"oh": "oh_no"},
            }
        }


app = FastAPI()


@app.get("/verify", response_model=UnionChild)
async def something(some: int):
    pass
```

**Comment by gunjeetgulatigit** [comment]
thanks @Rushyanth111 
This way of displaying scheme is working

**Comment by gunjeetgulatigit** [comment]
Hi @Rushyanth111 

Is there any  way to show multiple response scheme at swagger UI through this manual example step .. ?

**Comment by ghost** [comment]
@gunjeetgulatigit Yes, the example is indicative of your needs, you should edit the schema, according to the Open API Documentations, as it is manually set. 


As a small side note, please close the issue if the help required is satisfactory. Thanks!

**Comment by gunjeetgulatigit** [comment]
Thanks @Rushyanth111

**Comment by tiangolo** [comment]
Thanks for the help here @Rushyanth111 ! :clap: :bow: 

Thanks for reporting back and closing the issue @gunjeetgulatigit  :+1:

**Comment by singhyadvendra** [comment]
try this it works
responses={status.HTTP_200_OK: {"model": Union[child1, child2, child3]}, status.HTTP_404_NOT_FOUND: {"model": NotFoundMessage}}
